include/odp/api/spec/dma.h

```diff
@@ -0,0 +1,522 @@
+ *
+ * @param param         Parameter structure to be initialized
+ */
+void odp_dma_param_init(odp_dma_param_t *param);
+
+/**
+ * Create a DMA session
+ *
+ * @param name		DMA session name or NULL. Maximum string length is ODP_DMA_NAME_LEN.
+ * @param param		DMA session parameters
+ *
+ * @retval DMA session handle on success
+ * @retval ODP_DMA_INVALID on failure
+ */
+odp_dma_t odp_dma_create(const char *name, const odp_dma_param_t *param);
+
+/**
+ * Destroy DMA session
+ *
+ * @param dma		DMA session to be destroyed
+ *
+ * @retval 0 on success
+ * @retval <0 on failure
+ */
+int odp_dma_destroy(odp_dma_t dma);
+
+/**
+ * Find a DMA session by name
+ *
+ * @param name          DMA session name
+ *
+ * @retval Handle of the first matching DMA session
+ * @retval ODP_DMA_INVALID  DMA session could not be found
+ */
+odp_dma_t odp_dma_lookup(const char *name);
+
+/**
+ * Initialize DMA transfer parameters
+ *
+ * Initialize an odp_dma_transfer_param_t to its default values.
+ *
+ * @param param         Parameter structure to be initialized
+ */
+void odp_dma_transfer_param_init(odp_dma_transfer_param_t *param);
+
+/**
+ * Initialize DMA transfer completion parameters
+ *
+ * Initialize an odp_dma_compl_param_t to its default values.
+ *
+ * @param param         Parameter structure to be initialized
+ */
+void odp_dma_compl_param_init(odp_dma_compl_param_t *param);
+
+/**
+ * Synchronous DMA transfer operation
+ *
+ * Transfers data according to the parameters and returns when the transfer is complete.
+ *
+ * @param dma           DMA session
+ * @param transfer      Transfer parameters
+ *
+ * @retval 0 on success
+ * @retval <0 on failure
+ */
+int odp_dma_transfer_sync(odp_dma_t dma, const odp_dma_transfer_param_t *transfer);
+
+/**
+ * Start DMA transfer
+ *
+ * Start asynchronous DMA transfer operation. Transfer parameters specify data to be transferred.
+ * Completion parameters specify how transfer completion is reported.
+ *
+ * @param         dma           DMA session
+ * @param         transfer      Transfer parameters
+ * @param[in,out] completion    Transfer completion parameters
+ *
+ * @retval 0 on success
+ * @retval <0 on failure
+ */
+int odp_dma_transfer_start(odp_dma_t dma, const odp_dma_transfer_param_t *transfer,
+			   odp_dma_compl_param_t *completion);
+
+/**
+ * Start multiple DMA transfers
+ *
+ * Like odp_dma_transfer_start(), but starts 'num' transfers.
+ *
+ * @param         dma           DMA session
+ * @param         transfer      Array of transfer parameters
+ * @param[in,out] completion    Array of transfer completion parameters
+ * @param         num           Number of transfers to start. Both parameter arrays have this
+ *                              many elements.
+ *
+ * @retval Number of transfers started successfully (0 ... num)
+ * @retval <0 on failure
+ */
+int odp_dma_transfer_start_multi(odp_dma_t dma, const odp_dma_transfer_param_t *transfer[],
+				 odp_dma_compl_param_t *completion[], int num);
```

Comment:
transfer_order_t selects ordering, and NO_ORDER is the default. That allows implementation to process transfers in any order.

a) To keep API simple, I did not put segment order option into transfer_order_t. It can be added later, if we have an application that needs it, and application would not want to create separate transfers for those. Segment order is even more  limiting for performance than transfer order (e.g. vector instructions could target only one segment at a time, not separate segments).

b) This is covered now (within a session)

I'd say that typically application is interest only about when the entire transfer (or a set of transfers) has completed. ORDER_TRANSFER_COMPL supports that with minimal performance impact. ORDER_TRANSFER is more strict and supports application to request e.g. DMA to complete data transfer (A) first and only after that process another transfer (b) that turns a flag in memory (A is visible to others before B).